File: C++/source/Framework/Core/FrameworkSequenceGenerator.cpp
/***************************************************************************
 *   Copyright (C) 2005-2006 Vidyasagara Guntaka                                *
 *   All rights reserved.                                                  *
 *   Author : Vidyasagara Reddy Guntaka                                    *
 ***************************************************************************/

#include "Framework/Core/FrameworkSequenceGenerator.h"

namespace WaveNs
{

FrameworkSequenceGenerator::FrameworkSequenceGenerator ()
{
}

FrameworkSequenceGenerator::~FrameworkSequenceGenerator ()
{
}

void FrameworkSequenceGenerator::getInitializeSequenceDuringPrePhase (vector<WaveServiceId> &serviceIds) const
{
    serviceIds = m_waveServiceIdsToInitializeDuringPrePhase;
}

void FrameworkSequenceGenerator::getEnableSequenceDuringPrePhase (vector<WaveServiceId> &serviceIds) const
{
    serviceIds = m_waveServiceIdsToEnableDuringPrePhase;
}

void FrameworkSequenceGenerator::getInstallSequenceDuringPrePhase (vector<WaveServiceId> &serviceIds) const
{
    serviceIds = m_waveServiceIdsToInstallDuringPrePhase;
}

void FrameworkSequenceGenerator::getBootSequenceDuringPrePhase (vector<WaveServiceId> &serviceIds) const
{
    serviceIds = m_waveServiceIdsToBootDuringPrePhase;
}

void FrameworkSequenceGenerator::getShutdownSequenceDuringPostPhase (vector<WaveServiceId> &serviceIds) const
{
    serviceIds = m_waveServiceIdsToShutdownDuringPostPhase;
}

void FrameworkSequenceGenerator::getUninstallSequenceDuringPostPhase (vector<WaveServiceId> &serviceIds) const
{
    serviceIds = m_waveServiceIdsToUninstallDuringPostPhase;
}

void FrameworkSequenceGenerator::getUninitializeSequenceDuringPostPhase (vector<WaveServiceId> &serviceIds) const
{
    serviceIds = m_waveServiceIdsToUninitializeDuringPostPhase;
}

void FrameworkSequenceGenerator::getDisableSequenceDuringPostPhase (vector<WaveServiceId> &serviceIds) const
{
    serviceIds = m_waveServiceIdsToDisableDuringPostPhase;
}

void FrameworkSequenceGenerator::getInitializeSequence (vector<WaveServiceId> &serviceIds) const
{
    serviceIds = m_waveServiceIdsToInitialize;
}

void FrameworkSequenceGenerator::getEnableSequence (vector<WaveServiceId> &serviceIds) const
{
    serviceIds = m_waveServiceIdsToEnable;
}

void FrameworkSequenceGenerator::getInstallSequence (vector<WaveServiceId> &serviceIds) const
{
    serviceIds = m_waveServiceIdsToInstall;
}

void FrameworkSequenceGenerator::getBootSequence (vector<WaveServiceId> &serviceIds) const
{
    serviceIds = m_waveServiceIdsToBoot;
}

void FrameworkSequenceGenerator::getShutdownSequence (vector<WaveServiceId> &serviceIds) const
{
    serviceIds = m_waveServiceIdsToShutdown;
}

void FrameworkSequenceGenerator::getUninstallSequence (vector<WaveServiceId> &serviceIds) const
{
    serviceIds = m_waveServiceIdsToUninstall;
}

void FrameworkSequenceGenerator::getUninitializeSequence (vector<WaveServiceId> &serviceIds) const
{
    serviceIds = m_waveServiceIdsToUninitialize;
}

void FrameworkSequenceGenerator::getDisableSequence (vector<WaveServiceId> &serviceIds) const
{
    serviceIds = m_waveServiceIdsToDisable;
}

void FrameworkSequenceGenerator::getFailoverSequence (vector<WaveServiceId> &serviceIds) const
{
    serviceIds = m_waveServiceIdsToFailover;
}

void FrameworkSequenceGenerator::getZeroizeSequence (vector<WaveServiceId> &serviceIds) const
{
    serviceIds = m_waveServiceIdsToZeroize;
}

void FrameworkSequenceGenerator::addWaveServiceIdToAll (const WaveServiceId &waveServiceId, const bool &isForNormalPhase)
{
    if (true == isForNormalPhase)
    {
        m_waveServiceIdsToInitialize.push_back (waveServiceId);
        m_waveServiceIdsToEnable.push_back (waveServiceId);
        m_waveServiceIdsToInstall.push_back (waveServiceId);
        m_waveServiceIdsToBoot.push_back (waveServiceId);
        m_waveServiceIdsToShutdown.insert (m_waveServiceIdsToShutdown.begin (), waveServiceId);
        m_waveServiceIdsToUninstall.insert (m_waveServiceIdsToUninstall.begin (), waveServiceId);
        m_waveServiceIdsToUninitialize.insert (m_waveServiceIdsToUninitialize.begin (), waveServiceId);
        m_waveServiceIdsToDisable.insert (m_waveServiceIdsToDisable.begin (), waveServiceId);
    }
    else
    {
        m_waveServiceIdsToInitializeDuringPrePhase.push_back (waveServiceId);
        m_waveServiceIdsToEnableDuringPrePhase.push_back (waveServiceId);
        m_waveServiceIdsToInstallDuringPrePhase.push_back (waveServiceId);
        m_waveServiceIdsToBootDuringPrePhase.push_back (waveServiceId);
        m_waveServiceIdsToShutdownDuringPostPhase.insert (m_waveServiceIdsToShutdownDuringPostPhase.begin (), waveServiceId);
        m_waveServiceIdsToUninstallDuringPostPhase.insert (m_waveServiceIdsToUninstallDuringPostPhase.begin (), waveServiceId);
        m_waveServiceIdsToUninitializeDuringPostPhase.insert (m_waveServiceIdsToUninitializeDuringPostPhase.begin (), waveServiceId);
        m_waveServiceIdsToDisableDuringPostPhase.insert (m_waveServiceIdsToDisableDuringPostPhase.begin (), waveServiceId);
    }

    m_waveServiceIdsToFailover.push_back   (waveServiceId);
    m_waveServiceIdsToZeroize.push_back    (waveServiceId);
    
}

void FrameworkSequenceGenerator::addWaveServiceIdToInitialize (const WaveServiceId &waveServiceId, const bool &isForNormalPhase)
{
    m_waveServiceIdsToInitialize.push_back (waveServiceId);
}

void FrameworkSequenceGenerator::addWaveServiceIdToEnable (const WaveServiceId &waveServiceId, const bool &isForNormalPhase)
{
    m_waveServiceIdsToEnable.push_back (waveServiceId);
}

void FrameworkSequenceGenerator::addWaveServiceIdToInstall (const WaveServiceId &waveServiceId, const bool &isForNormalPhase)
{
    m_waveServiceIdsToInstall.push_back (waveServiceId);
}

void FrameworkSequenceGenerator::addWaveServiceIdToBoot (const WaveServiceId &waveServiceId, const bool &isForNormalPhase)
{
    m_waveServiceIdsToBoot.push_back (waveServiceId);
}

void FrameworkSequenceGenerator::addWaveServiceIdToShutdown (const WaveServiceId &waveServiceId, const bool &isForNormalPhase)
{
    m_waveServiceIdsToShutdown.push_back (waveServiceId);
}

void FrameworkSequenceGenerator::addWaveServiceIdToUninstall (const WaveServiceId &waveServiceId, const bool &isForNormalPhase)
{
    m_waveServiceIdsToUninstall.push_back (waveServiceId);
}

void FrameworkSequenceGenerator::addWaveServiceIdToUninitialize (const WaveServiceId &waveServiceId, const bool &isForNormalPhase)
{
    m_waveServiceIdsToUninitialize.push_back (waveServiceId);
}

void FrameworkSequenceGenerator::addWaveServiceIdToDisable (const WaveServiceId &waveServiceId, const bool &isForNormalPhase)
{
    m_waveServiceIdsToDisable.push_back (waveServiceId);
}

void FrameworkSequenceGenerator::addWaveServiceIdToFailover (const WaveServiceId &waveServiceId, const bool &isForNormalPhase)
{
    m_waveServiceIdsToFailover.push_back (waveServiceId);
}

void FrameworkSequenceGenerator::addWaveServiceIdToZeroize (const WaveServiceId &waveServiceId, const bool &isForNormalPhase)
{
    m_waveServiceIdsToZeroize.push_back(waveServiceId);
}

}

File: Java/source/com/CxWave/Wave/Framework/Core/FrameworkSequenceGenerator.java
/*************************************************************************************************************************
 * Copyright (C) 2015-2016 Vidyasagara Guntaka CxWave, Inc * All rights reserved. * Author : Vidyasagara Reddy Guntaka *
 *************************************************************************************************************************/

package com.CxWave.Wave.Framework.Core;

import java.util.Vector;

import com.CxWave.Wave.Framework.Type.WaveServiceId;

public class FrameworkSequenceGenerator
{
    private final Vector<WaveServiceId> m_waveServiceIdsToInitializeDuringPrePhase    = new Vector<WaveServiceId> ();
    private final Vector<WaveServiceId> m_waveServiceIdsToEnableDuringPrePhase        = new Vector<WaveServiceId> ();
    private final Vector<WaveServiceId> m_waveServiceIdsToInstallDuringPrePhase       = new Vector<WaveServiceId> ();
    private final Vector<WaveServiceId> m_waveServiceIdsToBootDuringPrePhase          = new Vector<WaveServiceId> ();

    private final Vector<WaveServiceId> m_waveServiceIdsToInitialize                  = new Vector<WaveServiceId> ();
    private final Vector<WaveServiceId> m_waveServiceIdsToEnable                      = new Vector<WaveServiceId> ();
    private final Vector<WaveServiceId> m_waveServiceIdsToInstall                     = new Vector<WaveServiceId> ();
    private final Vector<WaveServiceId> m_waveServiceIdsToBoot                        = new Vector<WaveServiceId> ();
    private final Vector<WaveServiceId> m_waveServiceIdsToShutdown                    = new Vector<WaveServiceId> ();
    private final Vector<WaveServiceId> m_waveServiceIdsToUninstall                   = new Vector<WaveServiceId> ();
    private final Vector<WaveServiceId> m_waveServiceIdsToUninitialize                = new Vector<WaveServiceId> ();
    private final Vector<WaveServiceId> m_waveServiceIdsToDisable                     = new Vector<WaveServiceId> ();

    private final Vector<WaveServiceId> m_waveServiceIdsToShutdownDuringPostPhase     = new Vector<WaveServiceId> ();
    private final Vector<WaveServiceId> m_waveServiceIdsToUninstallDuringPostPhase    = new Vector<WaveServiceId> ();
    private final Vector<WaveServiceId> m_waveServiceIdsToUninitializeDuringPostPhase = new Vector<WaveServiceId> ();
    private final Vector<WaveServiceId> m_waveServiceIdsToDisableDuringPostPhase      = new Vector<WaveServiceId> ();

    private final Vector<WaveServiceId> m_waveServiceIdsToFailover                    = new Vector<WaveServiceId> ();
    private final Vector<WaveServiceId> m_waveServiceIdsToZeroize                     = new Vector<WaveServiceId> ();

    public FrameworkSequenceGenerator ()
    {

    }

    public void getInitializeSequenceDuringPrePhase (final Vector<WaveServiceId> serviceIds)
    {
        serviceIds.addAll (m_waveServiceIdsToInitializeDuringPrePhase);
    }

    public void getEnableSequenceDuringPrePhase (final Vector<WaveServiceId> serviceIds)
    {
        serviceIds.addAll (m_waveServiceIdsToEnableDuringPrePhase);
    }

    public void getInstallSequenceDuringPrePhase (final Vector<WaveServiceId> serviceIds)
    {
        serviceIds.addAll (m_waveServiceIdsToInstallDuringPrePhase);
    }

    public void getBootSequenceDuringPrePhase (final Vector<WaveServiceId> serviceIds)
    {
        serviceIds.addAll (m_waveServiceIdsToBootDuringPrePhase);
    }

    public void getShutdownSequenceDuringPostPhase (final Vector<WaveServiceId> serviceIds)
    {
        serviceIds.addAll (m_waveServiceIdsToShutdownDuringPostPhase);
    }

    public void getUninstallSequenceDuringPostPhase (final Vector<WaveServiceId> serviceIds)
    {
        serviceIds.addAll (m_waveServiceIdsToUninstallDuringPostPhase);
    }

    public void getUninitializeSequenceDuringPostPhase (final Vector<WaveServiceId> serviceIds)
    {
        serviceIds.addAll (m_waveServiceIdsToUninitializeDuringPostPhase);
    }

    public void getDisableSequenceDuringPostPhase (final Vector<WaveServiceId> serviceIds)
    {
        serviceIds.addAll (m_waveServiceIdsToDisableDuringPostPhase);
    }

    public void getInitializeSequence (final Vector<WaveServiceId> serviceIds)
    {
        serviceIds.addAll (m_waveServiceIdsToInitialize);
    }

    public void getEnableSequence (final Vector<WaveServiceId> serviceIds)
    {
        serviceIds.addAll (m_waveServiceIdsToEnable);
    }

    public void getInstallSequence (final Vector<WaveServiceId> serviceIds)
    {
        serviceIds.addAll (m_waveServiceIdsToInstall);
    }

    public void getBootSequence (final Vector<WaveServiceId> serviceIds)
    {
        serviceIds.addAll (m_waveServiceIdsToBoot);
    }

    public void getShutdownSequence (final Vector<WaveServiceId> serviceIds)
    {
        serviceIds.addAll (m_waveServiceIdsToShutdown);
    }

    public void getUninstallSequence (final Vector<WaveServiceId> serviceIds)
    {
        serviceIds.addAll (m_waveServiceIdsToUninstall);
    }

    public void getUninitializeSequence (final Vector<WaveServiceId> serviceIds)
    {
        serviceIds.addAll (m_waveServiceIdsToUninitialize);
    }

    public void getDisableSequence (final Vector<WaveServiceId> serviceIds)
    {
        serviceIds.addAll (m_waveServiceIdsToDisable);
    }

    public void getFailoverSequence (final Vector<WaveServiceId> serviceIds)
    {
        serviceIds.addAll (m_waveServiceIdsToFailover);
    }

    public void getZeroizeSequence (final Vector<WaveServiceId> serviceIds)
    {
        serviceIds.addAll (m_waveServiceIdsToZeroize);
    }

    public void addWaveServiceIdToAll (final WaveServiceId waveServiceId)
    {
        addWaveServiceIdToAll (waveServiceId, true);
    }

    public void addWaveServiceIdToAll (final WaveServiceId waveServiceId, final boolean isForNormalPhase)
    {
        if (true == isForNormalPhase)
        {
            m_waveServiceIdsToInitialize.add (waveServiceId);
            m_waveServiceIdsToEnable.add (waveServiceId);
            m_waveServiceIdsToInstall.add (waveServiceId);
            m_waveServiceIdsToBoot.add (waveServiceId);
            m_waveServiceIdsToShutdown.add (0, waveServiceId);
            m_waveServiceIdsToUninstall.add (0, waveServiceId);
            m_waveServiceIdsToUninitialize.add (0, waveServiceId);
            m_waveServiceIdsToDisable.add (0, waveServiceId);
        }
        else
        {
            m_waveServiceIdsToInitializeDuringPrePhase.add (waveServiceId);
            m_waveServiceIdsToEnableDuringPrePhase.add (waveServiceId);
            m_waveServiceIdsToInstallDuringPrePhase.add (waveServiceId);
            m_waveServiceIdsToBootDuringPrePhase.add (waveServiceId);
            m_waveServiceIdsToShutdownDuringPostPhase.add (0, waveServiceId);
            m_waveServiceIdsToUninstallDuringPostPhase.add (0, waveServiceId);
            m_waveServiceIdsToUninitializeDuringPostPhase.add (0, waveServiceId);
            m_waveServiceIdsToDisableDuringPostPhase.add (0, waveServiceId);
        }

        m_waveServiceIdsToFailover.add (waveServiceId);
        m_waveServiceIdsToZeroize.add (waveServiceId);
    }

    public void addWaveServiceIdToInitialize (final WaveServiceId waveServiceId, final boolean isForNormalPhase)
    {
        m_waveServiceIdsToInitialize.add (waveServiceId);
    }

    public void addWaveServiceIdToEnable (final WaveServiceId waveServiceId, final boolean isForNormalPhase)
    {
        m_waveServiceIdsToEnable.add (waveServiceId);
    }

    public void addWaveServiceIdToInstall (final WaveServiceId waveServiceId, final boolean isForNormalPhase)
    {
        m_waveServiceIdsToInstall.add (waveServiceId);
    }

    public void addWaveServiceIdToBoot (final WaveServiceId waveServiceId, final boolean isForNormalPhase)
    {
        m_waveServiceIdsToBoot.add (waveServiceId);
    }

    public void addWaveServiceIdToShutdown (final WaveServiceId waveServiceId, final boolean isForNormalPhase)
    {
        m_waveServiceIdsToShutdown.add (waveServiceId);
    }

    public void addWaveServiceIdToUninstall (final WaveServiceId waveServiceId, final boolean isForNormalPhase)
    {
        m_waveServiceIdsToUninstall.add (waveServiceId);
    }

    public void addWaveServiceIdToUninitialize (final WaveServiceId waveServiceId, final boolean isForNormalPhase)
    {
        m_waveServiceIdsToUninitialize.add (waveServiceId);
    }

    public void addWaveServiceIdToDisable (final WaveServiceId waveServiceId, final boolean isForNormalPhase)
    {
        m_waveServiceIdsToDisable.add (waveServiceId);
    }

    public void addWaveServiceIdToFailover (final WaveServiceId waveServiceId, final boolean isForNormalPhase)
    {
        m_waveServiceIdsToFailover.add (waveServiceId);
    }

    public void addWaveServiceIdToZeroize (final WaveServiceId waveServiceId, final boolean isForNormalPhase)
    {
        m_waveServiceIdsToZeroize.add (waveServiceId);
    }
}
